gateway-js/src/index.ts

```diff
@@ -467,6 +445,102 @@ export class ApolloGateway implements GraphQLService {
     }
   }
 
+  private async updateCsdl(result: CsdlUpdate): Promise<void> {
+    if (result.id === this.compositionId) {
+      this.logger.debug('No change in composition since last check.');
+      return;
+    }
+
+    const previousSchema = this.schema;
+    const previousCsdl = this.parsedCsdl;
+    const previousCompositionId = this.compositionId;
+
+    if (previousSchema) {
+      this.logger.info('Updated CSDL was found.');
+    }
+
+    await this.maybePerformServiceHealthCheck(result);
+
+    this.compositionId = result.id;
+    this.parsedCsdl = parse(result.csdl);
+
+    if (this.queryPlanStore) this.queryPlanStore.flush();
+
+    const { schema, composedSdl } = this.createSchema({
+      csdl: result.csdl,
+    });
+
+    if (!composedSdl) {
+      this.logger.error(
+        "A valid schema couldn't be composed. Falling back to previous schema.",
+      );
+    } else {
+      this.schema = schema;
+      this.queryPlannerPointer = getQueryPlanner(composedSdl);
+
+      // Notify the schema listeners of the updated schema
+      try {
+        this.onSchemaChangeListeners.forEach((listener) =>
+          listener(this.schema!),
+        );
+      } catch (e) {
+        this.logger.error(
+          "An error was thrown from an 'onSchemaChange' listener. " +
+            'The schema will still update: ' +
+            ((e && e.message) || e),
+        );
+      }
+
+      if (this.experimental_didUpdateComposition) {
+        this.experimental_didUpdateComposition(
+          {
+            compositionId: result.id,
+            csdl: result.csdl,
+            schema: this.schema,
+          },
+          previousCompositionId && previousCsdl && previousSchema
+            ? {
+                compositionId: previousCompositionId,
+                csdl: print(previousCsdl),
+                schema: previousSchema,
+              }
+            : undefined,
+        );
+      }
+    }
+  }
+
+  private async maybePerformServiceHealthCheck(update: CompositionUpdate) {
+    // Run service health checks before we commit and update the new schema.
+    // This is the last chance to bail out of a schema update.
+    if (this.config.serviceHealthCheck) {
+      const serviceList = isCsdlUpdate(update)
+        ? this.serviceListFromCsdl(parse(update.csdl))
```

Comment:
We're already parsing this in the caller; can we avoid doing this twice?

Reply:
Yeah, we should just use the `gql` tag in all places so the parse result is cached for us.
ebe6be1

Reply:
My understanding is that `graphql-tag`'s cache is unbounded, so if a gateway runs for a long time loading different schemas, this will be a memory leak? (https://github.com/apollographql/graphql-tag/issues/356)
Probably we can manage to solve the specific case of the duplicate parse here without needing to rely on the global cache? Or we can reparse for now, which I think is better than the memory leak.